HDDS-3955. Unable to list intermediate paths on keys created using S3G.

bharatviswa504 opened the pull request:
## What changes were proposed in this pull request?

Keys created via the S3 Gateway currently use the createKey OM API to create the ozone key. Hence, when using a hdfs client to list intermediate directories in the key, OM returns key not found error. This was encountered while using fluentd to write Hive logs to Ozone via the s3 gateway.

Added a new config in OM to control the behavior of key create to create intermediate directories or not. In this way, keys created via ozone S3G can be accessed through fs API.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-3955

## How was this patch tested?

Added UT and IT.

arp7 commented on hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/OMConfigKeys.java:
```diff
@@ -239,4 +239,11 @@ private OMConfigKeys() {
           false;
+
+  // This config needs to be enabled, when S3G created objects will be used
+  // FileSystem.
```
used via FileSystem API.

bharatviswa504 replied:
done

arp7 commented:
Can you update the PR template a bit to describe the high-level approach (was fix made in OM or S3G)? 🙂

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:
```diff
@@ -221,8 +233,45 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+
+        // Check if a file or directory exists with same key name.
+        if (omDirectoryResult == DIRECTORY_EXISTS) {
+          throw new OMException("Can not write to directory: " + keyName,
```
Let's make the exception message slightly more descriptive to explain what is really going on. We can mention that createIntermediateDirs behavior is enabled and hence `/` has special interpretation.

bharatviswa504 replied:
Done

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:
```diff
@@ -221,8 +233,45 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager,
+            .getAllParentInfo(ozoneManager, keyArgs,
+                pathInfo.getMissingParents(), inheritAcls, trxnLogIndex);
+
+        // Add cache entries for the prefix directories.
```
@bharatviswa504 , one q. We are adding cache entries right away instead of at the end on success. Could this be a potential issue on the failure path?

bharatviswa504 replied:
Yes, there will be some intermediate directories left in DB, key commit failed. There will be intermediate directories created.
But I don't see that causing an issue. But cleaner thing is cleaning up entries. 
But this is a common issue for FileCreate and KeyCreate.

And in the case lets say key create itself failed in DB flush, we terminate DB.

arp7 commented:
The approach looks really good. It is surprisingly concise and elegant because I was expecting it to be a lot more code! Added a few review comments.

mukul1987 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:
```diff
@@ -328,6 +342,73 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
+    // Check intermediate paths are created
+    keyPath = keyPath.getParent();
+    while(keyPath != null) {
+      Assert.assertNotNull(omMetadataManager.getKeyTable().get(
```
Can we change this to fs.getFileStatus(keypath) so that returns status.isDir() ?

bharatviswa504 replied:
Added test

avijayanhwx commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:
```diff
@@ -265,4 +272,24 @@ public AuditMessage buildAuditMessage(AuditAction op,
+            .normalize().getPath();
+      }
+      if (!keyName.equals(normalizedKeyName)) {
+        LOG.debug("Normalized key {} to {} ", keyName, normalizedKeyName);
```
Can we print normalizedKeyName.substring(1) given we return that?

bharatviswa504 replied:
Done

avijayanhwx commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:
```diff
@@ -265,4 +272,24 @@ public AuditMessage buildAuditMessage(AuditAction op,
   }
+
+  @SuppressFBWarnings("DMI_HARDCODED_ABSOLUTE_FILENAME")
+  public static String getNormalizedKey(boolean enableFileSystemPaths,
```
What about moving the 'enableFileSystemPaths' flag outside, and call the method conditionally? In that way, the getNormalizedKey will stay a simple util function without knowing about application logic.

bharatviswa504 replied:
That requires in all over the code, we need to check this flag and call this. I feel this is better way.
If needed I can move the code of actual Normalize, to another utility if you expect this to be used else where.

bharatviswa504 replied:
Moved common logic to a new method.

avijayanhwx commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java:
```diff
@@ -0,0 +1,69 @@
+        getNormalizedKey(true, "////a/b/////c/...../d"));
+    Assert.assertEquals("a/b/d",
+        getNormalizedKey(true, "/a/b/c/../d"));
+    Assert.assertEquals("a/d",
```
Nit. Can we add a test for key with length = 1? Since we return .substring(1) in the method, trying to be cautious here.

bharatviswa504 replied:
Done

arp7 commented on hadoop-hdds/common/src/main/resources/ozone-default.xml:
```diff
@@ -2478,4 +2478,18 @@
+    <name>ozone.om.enable.filesystem.paths</name>
+    <tag>OZONE, OM</tag>
+    <value>false</value>
+    <description>If true, OM will normalize the key paths and also create
```
I would rephrase first sentence a bit to state that key names will be interpreted as file system paths. `/` will be treated as a special character and paths will be normalized and must follow Unix filesystem path naming conventions.

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:
```diff
@@ -265,4 +272,30 @@ public AuditMessage buildAuditMessage(AuditAction op,
+          .normalize().getPath();
+    }
+    if (!keyName.equals(normalizedKeyName)) {
+      LOG.debug("Normalized key {} to {} ", keyName,
```
Thanks for adding this log message. It will be super useful!

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:
```diff
@@ -79,15 +79,11 @@ public static OMPathInfo verifyFilesInPath(
-          result = OMDirectoryResult.FILE_EXISTS_IN_GIVENPATH;
-        }
-      } else if (omMetadataManager.getKeyTable().isExist(dbDirKeyName)) {
+      // Check first for dir. This is to handle leading "/" in path, which
```
Leading `/` in path should be dropped at this point, right?

bharatviswa504 replied:
Yes. This change was required, when we are allowing leading / in keyName. I will revert the change

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileRequest.java:
```diff
@@ -79,15 +79,11 @@ public static OMPathInfo verifyFilesInPath(
       String dbDirKeyName = omMetadataManager.getOzoneDirKey(volumeName,
           bucketName, pathName);
 
-      if (omMetadataManager.getKeyTable().isExist(dbKeyName)) {
```
I am a little bit worried about changes to OMFileRequest. There is a risk of changing the behavior and invalidating the app-compat testing we have done so far.

bharatviswa504 replied:
This change can be reverted as now we don't allow leading / in path names when OMConfig enable.filesystems is set to true. Will updated in my next PR update.

bharatviswa504 commented:
Thank You @arp7 and @avijayanhwx for the review.
I have addressed review comments.

bharatviswa504 commented:
Rebased with latest apache main branch.

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/OMClientRequest.java:
```diff
@@ -265,4 +273,72 @@ public AuditMessage buildAuditMessage(AuditAction op,
+          break;
+        }
+
+        // The string may end with a /, but not have
```
Does this mean we are allowing trailing `/`?

arp7 replied:
I realized this is checked later in `OmKeyCreateRequest`.

bharatviswa504 replied:
Yes, we don't allow for a keyName with trailing /. This method can return after normalization in some cases. So, we checked in OmKeyCreateRequest

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/file/OMFileCreateRequest.java:
```diff
@@ -141,7 +141,6 @@ public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
               ozoneManager.getPreallocateBlocksMax(),
               ozoneManager.isGrpcBlockTokenEnabled(),
               ozoneManager.getOMNodeId());
-
```
Spurious change?

bharatviswa504 replied:
Removed it

arp7 commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/OMKeyCreateRequest.java:
```diff
@@ -91,6 +99,19 @@ public OMRequest preExecute(OzoneManager ozoneManager) throws IOException {
+          ozoneManager.getEnableFileSystemPaths(), keyPath);
+      if (keyPath.endsWith("/")) {
+        throw new OMException("Invalid KeyPath: " + keyPath,
+            OMException.ResultCodes.INVALID_KEY_NAME);
```
Let's also mention why it is invalid, else user may be scratching their head. 🙂

bharatviswa504 replied:
Done

arp7 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
+    keyName = "././r1/r2/";
+    createAndCheck(keyName);
+
+    // converted to ..d1/d2/d3
```
Why is this converted to `../d1/d2/d3`? Should it be failed instead due to trailing `/`?

bharatviswa504 replied:
..d1/d2/d3/ -> ../d1/d2/d3 it removed trailing /. Normalize will take care of removing additional /

bharatviswa504 replied:
we will fail, even after normalize if we have a case like /a/. -> a/

arp7 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
+    keyName = "/////";
+    checkNotAValidPath(keyName);
+
+    keyName = "../../b/c";
```
Also let's add a check for `../a/b`.

bharatviswa504 replied:
Done

arp7 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
+
+  private void checkCreatedPaths(OMKeyCreateRequest omKeyCreateRequest,
+      OMRequest omRequest, String keyName) throws Exception {
+    keyName = omKeyCreateRequest.validateAndNormalizeKey(true, keyName);
```
Let's add a separate test case for `validateAndNormalizeKey` itself. All it does is call `validateAndNormalizeKey` with bunch of hard-coded input and asserts on the expected output.

arp7 replied:
Ignore this comment, I see it above now. Was reviewing files out of order.

arp7 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/key/TestOMKeyCreateRequest.java:
```diff
@@ -327,7 +344,164 @@ private OMRequest createKeyRequest(boolean isMultipartKey, int partNumber) {
+    keyName = omKeyCreateRequest.validateAndNormalizeKey(true, keyName);
+    // Check intermediate directories created or not.
+    Path keyPath = Paths.get(keyName);
+    checkIntermediatePaths(keyPath);
```
Nice!

arp7 commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/request/TestNormalizePaths.java:
```diff
@@ -0,0 +1,109 @@
+    Assert.assertEquals("a/b/c/d",
+        validateAndNormalizeKey(true, "////a/b/////c/d"));
+    Assert.assertEquals("a/b/c/...../d",
+        validateAndNormalizeKey(true, "////a/b/////c/...../d"));
```
Bharat, more than two dots should normalize to itself. In Unix I can create a dir named `...` or `.....`.

bharatviswa504 replied:
yes, you are right. It is the samething right, we are checking a/b/c/...../d, just it has removed additional and leading /

arp7 replied:
Right, I got confused.

bharatviswa504 commented:
> Few minor comments, and couple of questions on the expected normalizations.
> 
> Overall the change looks good. Thanks for adding comprehensive test cases.

Thank You @arp7 for the review. addressed review comments.
